Let validateaddress locate error in Bech32 address

meshcollider opened the pull request:
Addresses (partially) #16779 - no GUI change in this PR

Adds a LocateError function the bech32 library, which is then called by `validateaddress` RPC, (and then eventually from a GUI tool too, future work). I think modifying validateaddress is nicer than adding a separate RPC for this.
Includes tests.

Based on https://github.com/sipa/bech32/blob/master/ecc/javascript/bech32_ecc.js
Credit to sipa for that code

emilengler commented:
See #16779
Concept ACK

promag commented on src/bech32.cpp:
```diff
@@ -189,4 +376,96 @@ std::pair<std::string, data> Decode(const std::string& str) {
+        error = "Invalid character";
+        return ind;
+    }
+    if (str.size() > 90) {
```
9d91c4a046d990d7f648328dfed982ad38c87ff1

Should be first test?

promag commented on src/bech32.cpp:
```diff
@@ -189,4 +376,96 @@ std::pair<std::string, data> Decode(const std::string& str) {
+        int s0 = syn & 0x3FF;
+        int s1 = (syn >> 10) & 0x3FF;
+        int s2 = syn >> 20;
+        int l_s0 = GF1024_LOG[s0];
```
9d91c4a046d990d7f648328dfed982ad38c87ff1

Could add at the top:
```cpp
static_assert(sizeof(GF1024_LOG) / sizeof(GF1024_LOG[0]) == 2014, "GF1024_LOG length should be 1024");
```
Same for `GF1024_EXP`.

promag commented on src/bech32.h:
```diff
@@ -25,6 +25,9 @@ std::string Encode(const std::string& hrp, const std::vector<uint8_t>& values);
 std::pair<std::string, std::vector<uint8_t>> Decode(const std::string& str);
 
+/** Return the position of an error in a Bech32 string. If no errors are present, return -1. */
+int LocateError(const std::string& str, std::string& error);
```
9d91c4a046d990d7f648328dfed982ad38c87ff1

Could add some tests to `src/test/bech32_tests.cpp`?

promag commented on src/bech32.cpp:
```diff
@@ -189,4 +376,96 @@ std::pair<std::string, data> Decode(const std::string& str) {
+                return str.size() - p1 - (p1 >= values.size() ? 2 : 1);
+            }
+        }
+        for (int p1 = 0; p1 < length; p1++) {
```
9d91c4a046d990d7f648328dfed982ad38c87ff1

nit, `++p1`.

promag commented on src/rpc/misc.cpp:
```diff
@@ -39,6 +40,8 @@ static UniValue validateaddress(const JSONRPCRequest& request)
             "  \"iswitness\" : true|false,     (boolean) If the address is a witness address\n"
             "  \"witness_version\" : version   (numeric, optional) The version number of the witness program\n"
             "  \"witness_program\" : \"hex\"     (string, optional) The hex value of the witness program\n"
+            "  \"error\" : \"message\"           (string, optional) The error message if the address is an invalid bech32 address\n"
```
f877a9b4af7da6632430eff898b957107514e6de

Could include only when prefix matches `CChainParams::Bech32HRP`?

meshcollider replied:
I thought about this but I wasn't sure, because one of the checks is that the HRP is present and valid too. You're probably right, that would be better for the user.

promag reviewed (commented):
Concept ACK, light review, haven't compared with reference code.

sipa commented:
Would it be possible to add functions to key_io that provide this functionality for all addresses (with just basic error messages for anything but bech32)? That way the RPC code doesn't need to directly invoke the bech32 module.

promag commented:
+1 @sipa suggestion, something like
```cpp
CTxDestination DecodeDestination(const std::string& str, std::string& error);
CTxDestination DecodeDestination(const std::string& str)
{
    const std::string& error;
    return DecodeDestination(str, error);
}
```
And maybe also embed the error index in the message for now.

meshcollider commented:
I've added boost tests, fixed the above nits, and added a final commit with one implementation of the above idea. I can drop the final commit if it's not a good approach, but this is the cleanest way I could come up with

meshcollider commented:
Rebased

Sjors commented:
Concept ACK. I would prefer to make `address_type` an optional hint, and instead guess the address type.

Distinguishing between base58 and bech32 could be done by counting how many characters of the address are part of the base58/bech32 character set. You can then use `error_index` to indicate the position of a character that doesn't belong in the set.

luke-jr commented on src/rpc/misc.cpp:
```diff
@@ -28,6 +29,7 @@ static UniValue validateaddress(const JSONRPCRequest& request)
                 "\nReturn information about the given bitcoin address.\n",
                 {
                     {"address", RPCArg::Type::STR, RPCArg::Optional::NO, "The bitcoin address to validate"},
+                    {"address_type", RPCArg::Type::STR, RPCArg::Optional::OMITTED_NAMED_ARG, "The address type provided, used to detect errors. Options are \"legacy\", \"p2sh-segwit\", and \"bech32\"."},
```
Why can't this be detected from the input?

meshcollider replied:
Because if we assume it has errors in it, then we can't rely on checks like whether all the characters are bech32 or base58 or whatever. But I like Sjors' suggestion of counting the characters to find the most likely. I'll try that when I rebase this.

luke-jr replied:
FWIW, for Knots I am just checking for the Bech32HRP to match

Sjors replied:
That sounds simple enough.

luke-jr replied:
https://github.com/bitcoinknots/bitcoin/commit/65b45fa8f852a6d375061929e6c943c2afbbe681

luke-jr commented on test/functional/rpc_validateaddress.py:
```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env python3
```
File needs +x permission

luke-jr commented:
Once this is merged, I have the GUI changes ready to go...

gmaxwell commented:
Why is this limited to only returning a single error position?

gmaxwell commented:
@meshcollider  ?

sipa commented:
Concept ACK to adding Bech32 error location to `validateaddress`, but echoing earlier comments:
* Why only one error? The reference code in https://github.com/sipa/bech32 finds two, and I don't think just returning one makes things meaningfully simpler. Even if not implemented, the interface should probably return an array of error positions.
* It feels unnecessary to pass an address type. Just only run the bech32 error location code when the HRP matches (and treat as base58 otherwise).

meshcollider commented:
Fair points, I'll modify to return more than one error. Re. address type, we discussed it a while back in IRC, I can't remember if there was a reason we didn't base it solely on HRP or not, but I am happy to do it that way if you want because I can't see a real downside now.

adamjonas commented on test/functional/rpc_validateaddress.py:
```diff
@@ -0,0 +1,136 @@
+        address = "muehQnn2P5NhiRCGg9HewAWrXLRy2d7TZE"
+        res = self.nodes[0].validateaddress(address, "legacy")
+        assert_equal(res['isvalid'], False)
+        assert_equal(res['error'], "Invalid checksum for Base58 address")
```
This legacy address produces the error "Invalid Base58 character in address".

adamjonas commented on test/functional/rpc_validateaddress.py:
```diff
@@ -0,0 +1,136 @@
+        address = "1BvBMSEYstWetqTFn5Au4m4GFg7xJaNVN2"
+        res = self.nodes[0].validateaddress(address, "legacy")
+        assert_equal(res['isvalid'], False)
+        assert_equal(res['error'], "Invalid prefix for Base58 address")
```
Same here - this address produces the error "Invalid Base58 character in address".

adamjonas reviewed (commented):
A couple of touch-ups to the tests needed when you rebase.

Sjors commented:
Needs rebase

felipsoarez reviewed (commented):
Concept ACK, light review, haven't compared with reference code.

meshcollider committed:
Add release notes for validateaddress Bech32 error detection

meshcollider committed:
Add boost tests for bech32 error detection

meshcollider commented:
Rebased

meshcollider commented:
Rebased again on 21736 to fix CI

MarcoFalke commented:
From ci:

AssertionError: not(Invalid HRP or Base58 character in address == Invalid address format)

Sjors reviewed (commented):
3bc568d looks mostly good and works.

I don't fully understand the moon math involved in the error position detection, but it seems to work and it's run if and only if the regular decoding fails.

achow101 reviewed (commented):
I'm not sure that I agree with having separate functions for locating the errors that callers need to call. Rather I think it would be better to modify the decode functions themselves to locate the errors too when it fails to decode. This way we aren't duplicating code and we don't have to try to figure out where decode failed after the fact.

meshcollider committed:
More detailed error checking for base58 addresses

meshcollider committed:
Add Bech32 error location function

meshcollider committed:
Add error_locations to validateaddress RPC

meshcollider committed:
Refactor and add more tests for validateaddress

meshcollider commented:
Pushed a substantial rebase + rework. Addresses all comments and suggestions above. Notably this now makes DecodeAddress return the error locations rather than creating a separate function as @achow101 suggested.

achow101 commented:
ACK 1414005c5b24f13bb4da850c62c7c3cdc6436ad1

Reviewed everything except the actual bech32(m) error location code as I currently don't understand the math there. Also tested manually with a few addresses.

sipa commented on src/rpc/misc.cpp:
```diff
@@ -51,6 +51,10 @@ static RPCHelpMan validateaddress()
                         {RPCResult::Type::NUM, "witness_version", /* optional */ true, "The version number of the witness program"},
                         {RPCResult::Type::STR_HEX, "witness_program", /* optional */ true, "The hex value of the witness program"},
                         {RPCResult::Type::STR, "error", /* optional */ true, "Error message, if any"},
+                        {RPCResult::Type::ARR, "error_locations", "Indices of error locations in address, if known (e.g. Bech32 errors)",
```
Maybe say "likely error locations"; if you have too many errors, there is no telling where they are.

meshcollider commented:
Addressed @sipa's comments, thanks!

meshcollider committed:
Add lots of comments to Bech32

meshcollider committed:
Modify copyright header on Bech32 code

meshcollider commented:
Rebased 974227bb4576288795984085387ae0802fc3066e -> befad5eef168a193b142c4e6a1d0f120a85aae65 to fix conflict with #22794

meshcollider commented:
@gmaxwell you may also like to review 🙂

ryanofsky reviewed (approved):
Code review ACK 88cc4810926e4f5af6757ee1b0eed61abda3d746 with caveat that I only checked the new `LocateErrors` code to try to verify it didn't have unsafe or unexpected operations or loop forever or crash. Did not try to verify behavior corresponds to the spec. In the worst case bugs here should just affect error messages not actual decoding of addresses so this seemed ok.

I left various review suggestions, but feel to ignore them or just use ones that seem useful.

I do think it should be possible to add more test coverage. I think it would be nice to have a fuzz test that would encode a random address, decode it and verify the address was the same, introduce random errors, and verify the errors were detected at the right locations. Also maybe there could be unit tests or constexpr checks that verify the contents of the new GF1024_EXP / GF1024_LOG arrays.

w0xlt reviewed (approved):
tACK 88cc481

The command below returns the "error_locations" field on the PR branch.
```
./src/bitcoin-cli validateaddress bc1qctxsy54qjhk0dl5ea0pd2xtth2qq9n78wlk39m
```

Master branch:
```
{
  "isvalid": false,
  "error": "Invalid address format"
}
```

PR Branch:
```
{
  "isvalid": false,
  "error_locations": [
    29
  ],
  "error": "Invalid checksum"
}
```

laanwj commented on src/key_io.h:
```diff
@@ -23,7 +23,7 @@ std::string EncodeExtPubKey(const CExtPubKey& extpubkey);
 std::string EncodeDestination(const CTxDestination& dest);
 CTxDestination DecodeDestination(const std::string& str);
-CTxDestination DecodeDestination(const std::string& str, std::string& error_msg);
+CTxDestination DecodeDestination(const std::string& str, std::string& error_msg, std::vector<int>* error_locations = nullptr);
```
Agree with @ryanofsky here. References are hard to read as output arguments.
If you're going to make a separate function declaration, I'd prefer a function that returns it as part of the return value. E.g. as a pair or tuple.
But this is fine as-is.

laanwj commented:
Code review and manually tested ACK 88cc4810926e4f5af6757ee1b0eed61abda3d746
Thanks for adding all the bech32 comments.

sipa reviewed (commented):
These are some review comments I had queued up. I'll go through the rest still, and perhaps open a PR for any serious ones I still find.

meshcollider commented:
Thanks for the review comments! I will open a follow-up PR today.

ryanofsky commented on src/bech32.cpp:
```diff
@@ -227,4 +431,103 @@ DecodeResult Decode(const std::string& str) {
+            int s0 = syn & 0x3FF;
+            int s1 = (syn >> 10) & 0x3FF;
+            int s2 = syn >> 20;
+            int l_s0 = GF1024_LOG[s0];
```
re: https://github.com/bitcoin/bitcoin/pull/16807#discussion_r733208374

> @ryanofsky Alternatively, would it help if there was an (additional) `& 0x3FF` inside the `[]`?

I think most ideal thing would be for it to use something like [`std::array::at`](https://en.cppreference.com/w/cpp/container/array/at) and throw an exception in case of a bug. Alternative of adding `& 0x3FF` to lookups could help avoid undefined behavior, but also clutter the code and mask potential bugs. What's best just depends on what your priorities are. No strong opinions from me!

MarcoFalke commented on src/key_io.cpp:
```diff
@@ -98,15 +102,27 @@ CTxDestination DecodeDestination(const std::string& str, const CChainParams& par
-        // Set potential error message.
-        // This message may be changed if the address can also be interpreted as a Bech32 address.
-        error_str = "Invalid prefix for Base58-encoded address";
+        if (!std::equal(script_prefix.begin(), script_prefix.end(), data.begin()) &&
```
Error: attempt to subscript a past-the-end iterator 1 step from its current 
position, which falls outside its dereferenceable range.

https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=41364
